### PR TITLE
Better markdown support

### LINKED ISSUE #43: Properly handle mentions on metatags
We should do the same thing we are doing in the message content component but it should be converted to plain text

### PR DESCRIPTION
**Pull request by RiskyMH**
Closes #43 

This builds upon #50 and separates out the many functions involved to allow for a basic (for metatags) and still have the more advanced main message view. I made use of `unstable_cache` (it seemed trustworthy enough) to cache result of the mentions which should also allow for mentions to not need to be re-fetched in the same thread. I also decided to not keep your bulk query because it didn't seem like many of the same mention type would occur at the same time.

The next step with improving this is #23 and some other way that allows for the channel mentions to be a link.

### CODE REVIEW

**Comment by rafaelalmeidatk** [comment]
The bulk queries are mostly to reduce the amount of fired queries per number of overall mentions in the message, so if you mention 3 different users (which is more likely than the same user more than once), it will always only do 1 query instead of 3. 

You correctly identified the problem of duplicate queries because we need to parse the messages both for the metatags and again for the page, but I think the overall implementation of the PR could be simpler. My thoughts:
1. We can avoid defining our own regexes and continue relying on the dry pass of `discord-markdown` to get the user and channel mentions
2. `fetchMentions` was a good refactor, but we could've kept the old logic that was simpler and batched the mentions. You can wrap `fetchMentions` in `cache` (from React) to avoid running the parsing and querying logic twice in the page. We don't need to cache this data for other requests (using `unstable_cache`) because mentions aren't that prevailing in messages and the queries are very cheap

So TL;DR: it would be simpler to move the current logic into a function wrapped in `cache` that returns the `posts`, `users` and `channels` variables

**Comment by RiskyMH** [comment]
> The bulk queries are mostly to reduce the amount of fired queries per number of overall mentions in the message, so if you mention 3 different users (which is more likely than the same user more than once), it will always only do 1 query instead of 3.

I do kind of agree with this. However, personally I have only seen people ping someone that helped them to ask a clarifying question. And that the same thread can be including a different list of mentions (could be same mentions in a different order). But, mentions are kind of rare, so this level of optimization (every mention individually cached) could be excessive. 

> We can avoid defining our own regexes and continue relying on the dry pass of `discord-markdown` to get the user and channel mentions

I did this because it seemed a bit excessive to use a markdown parser to do a simple change. As well as making it easier to change libraries as that part is already done (at least the method of fetching mention if async). Another reasoning behind the custom regex was because it was being done for links, so I wanted to make it more consistent.

> `fetchMentions` was a good refactor, but we could've kept the old logic that was simpler and batched the mentions. You can wrap `fetchMentions` in `cache` (from React) to avoid running the parsing and querying logic twice in the page. We don't need to cache this data for other requests (using `unstable_cache`) because mentions aren't that prevailing in messages and the queries are very cheap

To be honest, I kind of forgot about react's cache. I chose to do the individual mention cache because it seems likely that the thread of messages could include the same mention. The main reasoning was improving speed because the more cached, the more that doesn't have to be done to render it. 

more reasoning behind `unstable_cache` and improvements to with current implementation: I do think the cache should be longer length and possible be invalided by the discord bot (like sitemap) because that also improves speed (cache lasts longer and still correct data = more posts can use it = less time wasting for result). This can also be applied to many other things in this site because of its size. The cache can improve google's crawling because it may do it in waves (I don't know if this happens though), and the faster the respond time, the more pages can be indexed (crawl budget).

p.s. im happy to do those suggestions, I just wanted to share my reasoning to you

**Comment by RiskyMH** [comment]
Also could react's `useMemo` be more efficient as it would cache the whole markdown search also.

**Comment by rafaelalmeidatk** [comment]
> I do kind of agree with this. However, personally I have only seen people ping someone that helped them to ask a clarifying question. And that the same thread can be including a different list of mentions (could be same mentions in a different order). But, mentions are kind of rare, so this level of optimization (every mention individually cached) could be excessive.

The different list of mentions is a great point, so I agree that we can keep the fetching functions separated as you did. I am just not too sure about using `unstable_cache` because there are still open issues about it not working correctly and it feels like a premature optimization. I am pretty sure using only `cache` will be enough to dedupe the DB queries even on lengthy threads since it is all in the same request. So I think the way to go here is just to wrap the individual functions for user/posts/channels with `cache` instead of `unstable_cache`

File: apps/web/utils/discord-markdown.ts [hunk 1]
```diff
@@ -3,77 +3,158 @@ import { load } from 'cheerio'
...
+const QUERY_REVALIDATE_TIME = 60 // 1 minute
+
+const fetchUser = (userId: string) => unstable_cache(async () => {
+  return (
+    await db
+      .selectFrom('users')
+      .select(['snowflakeId', 'username'])
+      .where('snowflakeId', '=', userId)
+      .executeTakeFirst()
+  ) || null
```
Comment (rafaelalmeidatk) [inline apps/web/utils/discord-markdown.ts:34]: `executeTakeFirst` can return `undefined` if the row wasn't found so we could use it instead of changing it to `null`. This suggestion applies to the other functions

File: apps/web/utils/discord-markdown.ts [hunk 2]
```diff
@@ -3,77 +3,158 @@ import { load } from 'cheerio'
...
+  const [posts, users, channels] = (await Promise.all([
+    Promise.all(Array.from(postIds).map(fetchPost)),
+    Promise.all(Array.from(memberIds).map(fetchUser)),
+    Promise.all(Array.from(channelIds).map(fetchChannel)),
+  ]))
+
+  // Filter out null values
+  const postsFiltered = posts.filter((p) => p) as PostCache[]
+  const usersFiltered = users.filter((u) => u) as UserCache[]
+  const channelsFiltered = channels.filter((c) => c) as ChannelCache[]
```
Comment (rafaelalmeidatk) [inline apps/web/utils/discord-markdown.ts:84]: We probably don't need this, it is fine to have the arrays with possible `undefined` values because we check for them anyway later on

File: apps/web/utils/discord-markdown.ts [hunk 3]
```diff
@@ -3,77 +3,158 @@ import { load } from 'cheerio'
...
-      user(node) {
-        memberIds.add(node.id)
-        return ''
-      },
-      channel(node) {
-        channelIds.add(node.id)
-        return ''
-      },
-    },
+export const parseDiscordMessageBasic = async (content: string) => {
```
Comment (rafaelalmeidatk) [inline apps/web/utils/discord-markdown.ts:103]: Can't we use `discord-markdown` here as well? Otherwise we will still have markdown stuff in the text like `**this**`. We can do a pass to convert to HTML then maybe call `.text()` from cheerio to extract the plain text from the document

File: apps/web/utils/discord-markdown.ts [hunk 4]
```diff
@@ -3,77 +3,158 @@ import { load } from 'cheerio'
...
+  // Replace internal links
+  content = internalLink(content, posts)
+
+  // Parse the content
   const html = toHTML(content, {
     discordCallback: {
       user: (node) => {
         const user = users.find((u) => u.snowflakeId === node.id)
-        const userName = sanitizeText(user?.username ?? 'Unknown User')
+        if (!user) return `<i>@Unknown User</i>`
```
Comment (rafaelalmeidatk) [inline apps/web/utils/discord-markdown.ts:144]: What is the reason to wrap only unknown users with `<i>`?

**Review by rafaelalmeidatk — COMMENTED** [review]
I left a few other suggestions besides the change of `unstable_cache`. lmk your thoughts about them

File: apps/web/utils/discord-markdown.ts (hunk 4 above)
Comment (RiskyMH) [reply apps/web/utils/discord-markdown.ts:144]: On discord, unknown menuones are italics, so I wanted to continue that.

File: apps/web/utils/discord-markdown.ts (hunk 3 above)
Comment (RiskyMH) [reply apps/web/utils/discord-markdown.ts:103]: I just assumed it was inefficient to do more parsing then necessary and then re-parse the HTML into test. As well as Google __could__ understand the markdown and simply ignore. This could however be good for links because the simple won't have the long link (with masked links) but a simple text. I think possibly using a library to remove [markdown syntax](https://www.npmjs.com/package/remove-markdown) might be a better solution even though it gets back the the issue of having the Discord mentions twice. 

But I'll do this as it will simplify the code base (don't need to over complicate every thing).

File: apps/web/utils/discord-markdown.ts (hunk 1 above)
Comment (RiskyMH) [reply apps/web/utils/discord-markdown.ts:34]: The reason for this was that the types get messed up later on:
* `internalLink` wants `{ snowflakeId: string; title: string; }[]` but is given `({ snowflakeId: string; title: string; } | undefined)[]`
* less of an issue: the mentions finding just needs an extra question mark `.find((u) => u?.`

But your right, it is a bit unnecessary...

File: apps/web/utils/discord-markdown.ts (hunk 2 above)
Comment (RiskyMH) [reply apps/web/utils/discord-markdown.ts:84]: Same as above, but the reason it needed `null` was because the filter didn't like filtering undefined variables.

**Comment by RiskyMH** [comment]
I did most of your suggestions and added cache to provide better efficiency of parsing the same message... 

Lmk if you still want me to remove the `unstable_cache` (or change back to original/ react cache there).

Also, how long does react's cache last (couldn't find it anywhere).

File: apps/web/utils/discord-markdown.ts (hunk 1 above)
Comment (RiskyMH) [reply apps/web/utils/discord-markdown.ts]: Wait, now I remember the main issue was when the mention can't be found (dev and normal mode have this):
```js
@nextjs-forum/web:dev: - error ..\node_modules\next\dist\server\web\spec-extension\unstable-cache.js (84:35) @ parse
@nextjs-forum/web:dev: - error SyntaxError: "undefined" is not valid JSON
@nextjs-forum/web:dev:     at JSON.parse (<anonymous>)
@nextjs-forum/web:dev:     at async Promise.all (index 1)
@nextjs-forum/web:dev:     at async Promise.all (index 2)
@nextjs-forum/web:dev:     at async eval (./utils/discord-markdown.ts:66:38)
@nextjs-forum/web:dev:     at async eval (./utils/discord-markdown.ts:79:40)
@nextjs-forum/web:dev:     at async Post (./app/post/[id]/page.tsx:163:23)
@nextjs-forum/web:dev: digest: "3742489035"
@nextjs-forum/web:dev: null
```

Comment (rafaelalmeidatk) [reply apps/web/utils/discord-markdown.ts]: Ahh I see, `unstable_cache` is serializing the data to JSON so we can't use `undefined`. In this case using `null` is fine!

**Review by rafaelalmeidatk — APPROVED** [review]
Sorry for taking too long with this, I appreciate these changes :pray: 
I still prefer that we replace `unstable_cache` with `cache` so we don't need to worry about invaldating this cache when the user changes their name/avatar. Other than that this all looks good to me

**Comment by rafaelalmeidatk** [comment]
> Also, how long does react's cache last (couldn't find it anywhere).

Only during this request. So if we have a long thread with multiple mentions for the same user we are only gonna query this user once, but for the next page render it will be fetched again. I think this is fine because these reads are very cheap and we don't need to worry about having stale cache data (like you mentioned when the user changes their name)

**Comment by RiskyMH** [comment]
I have changed back to react cache, i think i have added it to the necessary places (lmk if you want their places changed)

**Review by rafaelalmeidatk — APPROVED** [review]
Awesome, thank you so much :raised_hands:
